janus.c

```diff
@@ -3925,60 +3955,38 @@ gint main(int argc, char *argv[])
 		g_strfreev(list);
 		list = NULL;
 	}
-	/* What is the local public IP? */
-	JANUS_LOG(LOG_VERB, "Available interfaces:\n");
+	/* What is the local IP? */
+	JANUS_LOG(LOG_VERB, "Selecting local IP address...\n");
+	gboolean local_ip_set = FALSE;
 	item = janus_config_get_item_drilldown(config, "general", "interface");
 	if(item && item->value) {
 		JANUS_LOG(LOG_VERB, "  -- Will try to use %s\n", item->value);
-	}
-	struct ifaddrs *myaddrs, *ifa;
-	int status = getifaddrs(&myaddrs);
-	char *tmp = NULL;
-	if (status == 0) {
-		for (ifa = myaddrs; ifa != NULL; ifa = ifa->ifa_next) {
-			if(ifa->ifa_addr == NULL) {
-				continue;
-			}
-			if((ifa->ifa_flags & IFF_UP) == 0) {
-				continue;
-			}
-			/* Check the interface name first: we can ignore that as well */
-			if(ifa->ifa_name != NULL && janus_ice_is_ignored(ifa->ifa_name))
-				continue;
-			if(ifa->ifa_addr->sa_family == AF_INET) {
-				struct sockaddr_in *ip = (struct sockaddr_in *)(ifa->ifa_addr);
-				char buf[16];
-				if(inet_ntop(ifa->ifa_addr->sa_family, (void *)&(ip->sin_addr), buf, sizeof(buf)) == NULL) {
-					JANUS_LOG(LOG_ERR, "\t%s:\tinet_ntop failed!\n", ifa->ifa_name);
-				} else {
-					JANUS_LOG(LOG_VERB, "\t%s:\t%s\n", ifa->ifa_name, buf);
-					/* Check if this IP address is in the ignore list, now */
-					if(janus_ice_is_ignored(buf))
-						continue;
-					if(item && item->value && !strcasecmp(buf, item->value)) {
-						local_ip = strdup(buf);
-						if(local_ip == NULL) {
-							JANUS_LOG(LOG_FATAL, "Memory error!\n");
-							exit(1);
-						}
-					} else if(strcasecmp(buf, "127.0.0.1")) {	/* FIXME Check private IP addresses as well */
-						if(tmp == NULL)	/* FIXME Take note of the first IP we find, we'll use it as a backup */
-							tmp = strdup(buf);
-					}
-				}
-			}
-			/* TODO IPv6! */
-		}
-		freeifaddrs(myaddrs);
-	}
-	if(local_ip == NULL) {
-		if(tmp != NULL) {
-			local_ip = tmp;
+		/* FIXME: IPv6 is not supported yet */
+		struct in_addr addr;
+		if (inet_pton(AF_INET, item->value, &addr) != 1) {
+			JANUS_LOG(LOG_WARN, "Invalid local IP specified: %s, guessing the default...\n", item->value);
 		} else {
-			JANUS_LOG(LOG_WARN, "Couldn't find any address! using 127.0.0.1 as local IP... (which is NOT going to work out of your machine)\n");
-			local_ip = g_strdup("127.0.0.1");
+			/* Verify that we can actually bind to that address */
+			struct sockaddr_in addr;
+			int r;
+			int fd = socket(AF_INET, SOCK_DGRAM, 0);
+			if (fd == -1)
+				JANUS_LOG(LOG_WARN, "Error creating test socket, falling back to detecting IP address...\n");
```

Comment:
it seems a little awkward to let the code plow on when `fd == -1` (though I guess `r < 0` will trigger)